### DISCUSSION TITLE
Questions about the best practice of "local replica(source control)"

### DISCUSSION THREAD

**Discussion by e-zz** [overleaf-workshop/Overleaf-Workshop#100] [Q&A]
Hi all, 
I am a bit confused about the correct way to use local replica created by this plugin's.

1. What does "enable/disable the Local Replica temporarily" mean? 
The connection is enabled/disabled temporarily? 

2. What does the source control mean here? 
Is it equivalent to "creating a new local replica"? If that's the case, we have to do `git init` by hand and the source control part is actually never touched by the plugin, right?

3. I encountered some unknown fatal error which corrupted all the PDFs in my project (twice by the same problem I think). I wonder is there any "Never-Do" list to show us the safe zone of using the plugin?

The problem I met : 
![image](https://github.com/iamhyc/Overleaf-Workshop/assets/37920155/b15c6317-bead-4c8d-a7df-362070b16d3d)
So when the warning message kept popping up, something revoked my local changes made in the local replica and also modified most of the PDFs in that project (got turned into empty files, or deleted)
![image](https://github.com/iamhyc/Overleaf-Workshop/assets/37920155/80462a13-7084-41be-b93c-c0d6dd86e6f4)

Much appreciated  for any ideas.

---

Well, things get so much worse in just 10 mins. The tex file is always empty. Even after restoring the file with git, it's cleared out by the plugin. At the same time, the same warning messages I mentioned show up again.

---
But strangely enough, after reopening the replica, everything looks fine. Almost all files are back (except that two tex files are still empty) . 
Then, the error recurrs (before I can do anything).

---
And on overleaf, the history shows that a lot of files got deleted and created repeatedly
![image](https://github.com/iamhyc/Overleaf-Workshop/assets/37920155/409f2351-5b09-4761-ba25-13857a146bec)

**Comment by iamhyc** [comment]
Hi @e-zz , sorry for the confusion.

In short, "source control" means the "Local Replica" tries to synchronize changes of the source code on overleaf and local folder.

> 1. What does "enable/disable the Local Replica temporarily" mean?
> The connection is enabled/disabled temporarily?

It means the synchronization is disabled or enabled.
- If disabled, no changes will be uploaded/downloaded.
- If enabled again, it tries to: a) fetch the remote changes, b) diff with possible local changes, c) apply merged changes to local, d) apply diff changes to remote.

>   2. What does the source control mean here?
>      Is it equivalent to "creating a new local replica"? If that's the case, we have to do `git init` by hand and the source control part is actually never touched by the plugin, right?

"Local Replica" is just one implementation of "source control". If you are interested, we have the API design [here](https://github.com/iamhyc/Overleaf-Workshop/blob/master/src/scm/index.ts) to implement and add a new "source control manager".

As I mentioned in the beginning: "Local Replica" is a "source control manager", without version control like git, simply synchronize the changes (file add/delete/change) forward and backward between overleaf server and local folder.

>   3. I encountered some unknown fatal error which corrupted all the PDFs in my project (twice by the same problem I think). I wonder is there any "Never-Do" list to show us the safe zone of using the plugin?

> The problem I met : 
> 
> Much appreciated for any ideas.
> 
> Well, things get so much worse in just 10 mins. The tex file is always empty. Even after restoring the file with git, it's cleared out by the plugin. At the same time, the same warning messages I mentioned show up again.
> 
> But strangely enough, after reopening the replica, everything looks fine. Almost all files are back (except that two tex files are still empty) . Then, the error recurrs (before I can do anything).
> 
> And on overleaf, the history shows that a lot of files got deleted and created repeatedly

Then it must be a bug. I thought I have this totally resolved in the release of `v0.8.2`.
Please let me know if you are using the latest plugin (version `v0.9.1`).

If so, please help create a formal issue [here](https://github.com/iamhyc/Overleaf-Workshop/issues/new?assignees=&labels=bug&projects=&template=bug_report.md&title=) with detailed steps to reproduce your situation. Thanks for your help!

**Comment by e-zz** [comment]
Thanks for the all the info. 

Now I see. I was a bit confused because in a local replica, we could click the clound icon,
![image](https://github.com/iamhyc/Overleaf-Workshop/assets/37920155/887da660-4621-4964-86c5-1573a8567567)
and create another replica here.
![image](https://github.com/iamhyc/Overleaf-Workshop/assets/37920155/d741f853-07b7-4223-a230-adf9388a0854)

I thought it was **to create source control, like git, for the local replica**. But in fact it's not 😂. It's actually for creating another copy of the same project if I understand correctly.

Indeed I am using `v0.9.1`. And now I will keep using the plugin with caution. If I found a way to constantly reproduce the bug, I will report it in an issue. 

Thanks again for your work on the plugin. It's awesome!

**Comment by e-zz** [comment]
Just one last question. 
Is the local replica supposed to be reused ? Or it's just for one-time purpose?

I start over again and recreated the replica after deleting it locally. Now the problem seems gone. 
However, when I  closing it with `ctrl+k f` and start vscode from the directory of the replica folder, somehow the plugin can not properly recognize the replica
- The cloud icon is gone
   ![image](https://github.com/iamhyc/Overleaf-Workshop/assets/37920155/4a6399ef-879c-46e1-9c6f-7a670cddc8c4)
- The `.overleaf` folder is still in the folder, and I checked there's nothing wrong with the content of  `.overleaf/settings.json`.
-  I can still see the indicator for my collaborator's cursor (actually me on the website). But it freezes and never moves. Here is the indicator I mentioned
  ![image](https://github.com/iamhyc/Overleaf-Workshop/assets/37920155/ed5fb731-6cf9-46f1-8670-beba1366e3e6)

- Local changes are not synced to overleaf automatically. 
    - Manual syncing is impossible since here we lost our cloud icon. Even the "upload to overleaf" option in the Invisible mode is not accessible now. 
   ![image](https://github.com/iamhyc/Overleaf-Workshop/assets/37920155/22a77f8f-fff7-47dc-896d-dab41969942a)
- Also, it's impossible to open the replica again from plugin's Hosts view, because the path is somehow invalid
  ![image](https://github.com/iamhyc/Overleaf-Workshop/assets/37920155/a6593a2f-3ba1-4e7a-8aa3-97f893ed68fe)
  ![image](https://github.com/iamhyc/Overleaf-Workshop/assets/37920155/5fdbf8ea-8a02-4a24-a64a-9980f404f13f)
- The only way to enter that replica I found till now, is to recreate a new replica at the same place, then everything is fine in the replica just created. Thankfully the plugin doesn't overwrite the folder (maybe because of the `.overleaf` existent there?)

There must be sth wrong but I don't know how to pin down the cause. And if necessary I could open an issue and continue our discussion there. Just let me know. Thanks!

**Reply by iamhyc** [reply]
Yes, please help create an issue.

It seems like you are using the plugin over a Remote SSH / WSL connection, thus the local path parse failed with `[object Object]`.

Please also remember to attach the error logs in the issue. Thank you!
